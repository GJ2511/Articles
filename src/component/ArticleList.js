import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import cx from 'classnames';

import Loader from './Loader';

function ArticleList({ articles, loading, handleRowClick, handleFavCellClick }) {
    if (loading) {
        return <Loader />;
    }

    const getTags = (tags) => {
        return tags.map((tag, index) => {
            return (
                <span className="badge ml-1 badge-info" key={index}>
                    {tag}
                </span>
            );
        });
    };

    const getFavIcon = (favorites) => {
        if (favorites) {
            return (
                <svg
                    className="bi bi-heart-fill"
                    width="1em"
                    height="1em"
                    viewBox="0 0 16 16"
                    fill="currentColor"
                    xmlns="http://www.w3.org/2000/svg"
                >
                    <path
                        fillRule="evenodd"
                        d="M8 1.314C12.438-3.248 23.534 4.735 8 15-7.534 4.736 3.562-3.248 8 1.314z"
                    />
                </svg>
            );
        }

        return (
            <svg
                className="bi bi-heart"
                width="1em"
                height="1em"
                viewBox="0 0 16 16"
                fill="currentColor"
                xmlns="http://www.w3.org/2000/svg"
            >
                <path
                    fillRule="evenodd"
                    d="M8 2.748l-.717-.737C5.6.281 2.514.878 1.4 3.053c-.523 1.023-.641 2.5.314 4.385.92 1.815 2.834 3.989 6.286 6.357 3.452-2.368 5.365-4.542 6.286-6.357.955-1.886.838-3.362.314-4.385C13.486.878 10.4.28 8.717 2.01L8 2.748zM8 15C-7.333 4.868 3.279-3.04 7.824 1.143c.06.055.119.112.176.171a3.12 3.12 0 0 1 .176-.17C12.72-3.042 23.333 4.867 8 15z"
                />
            </svg>
        );
    };

    const tableBody = articles.map((article, index) => {
        const classes = cx('btn rounded-pill text-white', {
            'badge-secondary': !article.favorited,
            'badge-primary': article.favorited,
        });

        return (
            <tr key={index} className="d-flex hover-hand">
                <td className="col-2" onClick={() => handleRowClick(article.slug)}>
                    {article.title}
                </td>
                <td className="col-3" onClick={() => handleRowClick(article.slug)}>
                    {article.description}
                </td>
                <td className="col-2" onClick={() => handleRowClick(article.slug)}>
                    {article.author.username}
                </td>
                <td className="col-2" onClick={() => handleRowClick(article.slug)}>
                    {getTags(article.tagList)}
                </td>
                <td className="col-1" onClick={() => handleFavCellClick(article)}>
                    <button type="button" className={classes}>
                        {getFavIcon(article.favorited)}{' '}
                        <span className="badge badge-light">{article.favoritesCount}</span>
                    </button>
                </td>
                <td className="col-2" onClick={() => handleRowClick(article.slug)}>
                    {moment(article.createdAt).format('dddd, MMMM Do YYYY')}
                </td>
            </tr>
        );
    });

    return (
        <table className="table table-hover table-bordered article_list">
            <thead className="thead-dark">
                <tr className="d-flex">
                    <th className="col-2">Title</th>
                    <th className="col-3">Description</th>
                    <th className="col-2">Author</th>
                    <th className="col-2">Tags</th>
                    <th className="col-1">Favorites Count</th>
                    <th className="col-2">Created At</th>
                </tr>
            </thead>
            <tbody>{tableBody}</tbody>
        </table>
    );
}

ArticleList.propTypes = {
    articles: PropTypes.array.isRequired,
    handleFavCellClick: PropTypes.func.isRequired,
    handleRowClick: PropTypes.func.isRequired,
    loading: PropTypes.bool.isRequired,
};

export default ArticleList;
