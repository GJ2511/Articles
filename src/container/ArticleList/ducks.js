import { takeLatest, call, put, select } from 'redux-saga/effects';
import map from 'lodash/map';

import ArticleService from '../../services/articleService';
import AuthService from '../../services/authService';

const initialState = {
    loading: true,
    articles: [],
    currentPage: 1,
    totalCount: 0,
    error: false,
    myFavArticle: false,
    myArticle: false,
};

const limit = 20;

const PREFIX = 'ARTICLES_LIST';
const GET_ARTICLES_LIST_REQUESTED = `${PREFIX}//GET_ARTICLES_LIST_REQUESTED`;
const GET_ARTICLES_LIST_SUCCESS = `${PREFIX}//GET_ARTICLES_LIST_SUCCESS`;
const GET_ARTICLES_LIST_ERROR = `${PREFIX}//GET_ARTICLES_LIST_ERROR`;
const TOGGLE_FAVORITE = `${PREFIX}//TOGGLE_FAVORITE`;
const MARK_FAVORITE = `${PREFIX}//MARK_FAVORITE`;
const UNMARK_FAVORITE = `${PREFIX}//UNMARK_FAVORITE`;
const SET_CURRENT_PAGE = `${PREFIX}//SET_CURRENT_PAGE`;
const TOGGLE_MY_FAV_ARTICLE = `${PREFIX}//TOGGLE_MY_FAV_ARTICLE`;
const TOGGLE_MY_ARTICLE = `${PREFIX}//TOGGLE_MY_ARTICLE`;

const articleListReducer = (state = initialState, action = {}) => {
    const { type, payload } = action;

    switch (type) {
        case GET_ARTICLES_LIST_REQUESTED:
            return {
                ...state,
                loading: true,
            };
        case GET_ARTICLES_LIST_SUCCESS:
            return {
                ...state,
                loading: false,
                articles: payload.articles,
                totalCount: Math.floor(payload.articlesCount / limit),
                ...(payload.resetPagination && { currentPage: 1 }),
            };
        case GET_ARTICLES_LIST_ERROR:
            return {
                ...state,
                ...initialState,
                error: true,
            };
        case SET_CURRENT_PAGE: {
            return {
                ...state,
                currentPage: payload,
            };
        }
        case UNMARK_FAVORITE: {
            const { articles } = state;

            return {
                ...state,
                articles: map(articles, (article) =>
                    article.slug === payload
                        ? { ...article, favorited: false, favoritesCount: article.favoritesCount - 1 }
                        : article
                ),
            };
        }
        case MARK_FAVORITE: {
            const { articles } = state;

            return {
                ...state,
                articles: map(articles, (article) =>
                    article.slug === payload
                        ? { ...article, favorited: true, favoritesCount: article.favoritesCount + 1 }
                        : article
                ),
            };
        }
        case TOGGLE_MY_FAV_ARTICLE: {
            return {
                ...state,
                myFavArticle: payload,
            };
        }
        case TOGGLE_MY_ARTICLE: {
            return {
                ...state,
                myArticle: payload,
            };
        }
        default:
            return state;
    }
};

export const getArticlesRequested = (resetPagination) => ({
    type: GET_ARTICLES_LIST_REQUESTED,
    payload: resetPagination,
});
export const setCurrentPage = (page) => ({
    type: SET_CURRENT_PAGE,
    payload: page,
});
export const toggleFavoriteRequested = (payload) => ({
    type: TOGGLE_FAVORITE,
    payload,
});
export const toggleMyFavArticle = (payload) => ({
    type: TOGGLE_MY_FAV_ARTICLE,
    payload,
});
export const toggleMyArticle = (payload) => ({
    type: TOGGLE_MY_ARTICLE,
    payload,
});

export const getState = (state) => state;

function* getArticles({ payload }) {
    try {
        const loggedInUser = AuthService.getLoggedInUser();
        const { articleListReducer, favTagReducer } = yield select();
        const resetPage = payload && payload.resetPage ? true : false;

        const params = {
            limit,
            offset: resetPage ? 0 : (articleListReducer.currentPage - 1) * limit,
            ...(favTagReducer.selectedTag && { tag: favTagReducer.selectedTag }),
            ...(articleListReducer.myArticle && loggedInUser !== null && { author: loggedInUser.username }),
            ...(articleListReducer.myFavArticle && loggedInUser !== null && { favorited: loggedInUser.username }),
        };

        const response = yield call([ArticleService, 'getArticles'], params);

        if (response.errors) {
            yield put({ type: GET_ARTICLES_LIST_ERROR });
        } else {
            yield put({
                type: GET_ARTICLES_LIST_SUCCESS,
                payload: { ...response, ...(resetPage && { resetPagination: true }) },
            });
        }
    } catch (error) {
        yield put({ type: GET_ARTICLES_LIST_ERROR });
    }
}

function* toggleFavorite({ payload: { slug, favorited } }) {
    const service = favorited ? 'unmarkArticleFav' : 'markArticleFav';

    try {
        yield call([ArticleService, service], slug);

        if (favorited) {
            yield put({
                type: UNMARK_FAVORITE,
                payload: slug,
            });
        } else {
            yield put({
                type: MARK_FAVORITE,
                payload: slug,
            });
        }
    } catch (error) {
        yield put({ type: GET_ARTICLES_LIST_ERROR });
    }
}

export function* articleListSaga() {
    yield takeLatest(GET_ARTICLES_LIST_REQUESTED, getArticles);
    yield takeLatest(TOGGLE_FAVORITE, toggleFavorite);
}

export default articleListReducer;
