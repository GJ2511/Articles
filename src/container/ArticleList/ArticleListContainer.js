import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import ArtileList from '../../component/ArticleList';
import ErrorList from '../../component/ErrorList';
import Pagination from '../../component/Pagination';
import AddNewBtn from '../../component/AddNewBtn';
import FavTagContainer from '../FavTag/FavTagContainer';
import { getArticlesRequested, setCurrentPage } from './ducks';

import historyService from '../../services/historyService';

class ArticleListContainer extends Component {
    componentDidMount() {
        this.props.getArticlesRequested();
    }

    onFavTagClick = () => {
        this.props.getArticlesRequested({ resetPage: true });
    };

    onRowClick = (slug) => {
        historyService.forwardTo(`/article/${slug}`);
    };

    onPageClick = (page) => {
        this.props.setCurrentPage(page);
        this.props.getArticlesRequested();
    };

    render() {
        const { error, totalCount, currentPage, loading, ...rest } = this.props;

        return (
            <div className="row mt-5">
                <div className="col-sm-12 col-md-3 col-lg-2 col-xl-2">
                    <AddNewBtn />
                    <FavTagContainer handleTagClick={this.onFavTagClick} />
                </div>
                <div className="col-sm-12 col-md-9 col-lg-10 col-xl-10">
                    {error ? (
                        <ErrorList errors={{ ERROR: ['Something Went Wrong'] }} />
                    ) : (
                        <>
                            <ArtileList {...rest} handleRowClick={this.onRowClick} loading={loading} />
                            <Pagination
                                totalCount={totalCount}
                                currentPage={currentPage}
                                loading={loading}
                                handlePageClick={this.onPageClick}
                            />
                        </>
                    )}
                </div>
            </div>
        );
    }
}

ArticleListContainer.propTypes = {
    articles: PropTypes.array.isRequired,
    currentPage: PropTypes.number.isRequired,
    error: PropTypes.bool.isRequired,
    getArticlesRequested: PropTypes.func.isRequired,
    loading: PropTypes.bool.isRequired,
    setCurrentPage: PropTypes.func.isRequired,
    totalCount: PropTypes.number.isRequired,
};

const mapStateToProps = ({ articleListReducer }) => ({
    articles: articleListReducer.articles,
    currentPage: articleListReducer.currentPage,
    error: articleListReducer.error,
    loading: articleListReducer.loading,
    totalCount: articleListReducer.totalCount,
});

export default connect(mapStateToProps, { getArticlesRequested, setCurrentPage })(ArticleListContainer);
