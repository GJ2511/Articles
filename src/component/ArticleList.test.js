import React from 'react';

import ArticleList from './ArticleList';

const props = {
    articles: [
        {
            slug: 'how-to-train-your-dragon',
            title: 'How to train your dragon',
            description: 'Ever wonder how?',
            body: 'It takes a Jacobian',
            tagList: ['dragons', 'training'],
            createdAt: '2016-02-18T03:22:56.637Z',
            updatedAt: '2016-02-18T03:48:35.824Z',
            favorited: false,
            favoritesCount: 0,
            author: {
                username: 'jake',
                bio: 'I work at statefarm',
                image: 'https://i.stack.imgur.com/xHWG8.jpg',
                following: false,
            },
        },
    ],
    handleFavCellClick: jest.fn(),
    handleRowClick: jest.fn,
    loading: false,
};

describe('<ArticleList tests />', () => {
    test('match snapshot', () => {
        expect(shallow(<ArticleList {...props} />)).toMatchSnapshot();
    });
});
