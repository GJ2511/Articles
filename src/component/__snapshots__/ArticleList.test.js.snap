// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArticleList tests /> match snapshot 1`] = `
<table
  className="table table-hover table-bordered article_list"
>
  <thead
    className="thead-dark"
  >
    <tr
      className="d-flex"
    >
      <th
        className="col-2"
      >
        Title
      </th>
      <th
        className="col-3"
      >
        Description
      </th>
      <th
        className="col-2"
      >
        Author
      </th>
      <th
        className="col-2"
      >
        Tags
      </th>
      <th
        className="col-1"
      >
        Favorite Count
      </th>
      <th
        className="col-2"
      >
        Created At
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      className="d-flex hover-hand"
      key="0"
    >
      <td
        className="col-2"
        onClick={[Function]}
      >
        How to train your dragon
      </td>
      <td
        className="col-3"
        onClick={[Function]}
      >
        Ever wonder how?
      </td>
      <td
        className="col-2"
        onClick={[Function]}
      >
        jake
      </td>
      <td
        className="col-2"
        onClick={[Function]}
      >
        <span
          className="badge ml-1 badge-info"
          key="0"
        >
          dragons
        </span>
        <span
          className="badge ml-1 badge-info"
          key="1"
        >
          training
        </span>
      </td>
      <td
        className="col-1"
        onClick={[Function]}
      >
        <button
          className="btn rounded-pill text-white badge-secondary"
          type="button"
        >
          <svg
            className="bi bi-heart"
            fill="currentColor"
            height="1em"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 2.748l-.717-.737C5.6.281 2.514.878 1.4 3.053c-.523 1.023-.641 2.5.314 4.385.92 1.815 2.834 3.989 6.286 6.357 3.452-2.368 5.365-4.542 6.286-6.357.955-1.886.838-3.362.314-4.385C13.486.878 10.4.28 8.717 2.01L8 2.748zM8 15C-7.333 4.868 3.279-3.04 7.824 1.143c.06.055.119.112.176.171a3.12 3.12 0 0 1 .176-.17C12.72-3.042 23.333 4.867 8 15z"
              fillRule="evenodd"
            />
          </svg>
           
          <span
            className="badge badge-light"
          >
            0
          </span>
        </button>
      </td>
      <td
        className="col-2"
        onClick={[Function]}
      >
        Thursday, February 18th 2016
      </td>
    </tr>
  </tbody>
</table>
`;
